openedx/openedx-platform
# feat: Expose xmodule xblocks Sass variables as vanilla CSS variables

Ticket: https://github.com/openedx/edx-platform/issues/35173

## Following script is used to extract out the required sass variables to expose:** 
```
# Search for all Sass var references, sort them, and save them to a file using ripgrep
rg '\$[a-zA-Z0-9-_]*'   xmodule/assets/**/*scss --no-line-number --no-filename --only-matching \
    | sort | uniq > variable_references.txt

# Search for local declarations of Sass vars (like '$selected_color: ...'), sort them,
# and save them to a file.
rg '\$[a-zA-Z0-9-_]*: ' xmodule/assets/**/*scss --no-line-number --no-filename --only-matching \
    | sed 's/..$//' | sort | uniq > local_variable_declarations.txt

# Compare the set of var references with the set of local declarations.
# Variables which are referenced but *don't* have local declarations must be themeable Sass variable.
# Save them to a file.
diff --unified local_variable_declarations.txt variable_references.txt \
    | rg '^\+' | sed 's/^.//'  > themable_variables.txt
```

## How to test:
### Test in xmodule xblocks:
1. Checkout the PR branch
2. Run `npm run compile-sass` or `npm run compile-sass-dev` in lms shell. Restart lms if changes don't reflect.
3. Open any XBlock in the lms or cms. Open dev tools window and find the css file in the scripts window. Select the xblock first with the Inspect selector arrow.
4. We can also hit following command in console to verify the variable
```shell
getComputedStyle(document.documentElement).getPropertyValue('--baseline')
```

### Test in extracted xblocks:
1. Checkout the extracted word-cloud xblocks [PR](https://github.com/openedx/edx-platform/pull/35140). Follow instructions on the PR.
2. Convert the word-cloud xblock hard code variable into css global variable in [file](https://github.com/openedx/edx-platform/blob/127c5c1ce276b3e0bf16dc5f1f65bb48b1306687/xmodule/assets/word_cloud/_display.scss#L6-L8) as shared below
```css
.input-cloud {
    margin: var(--baseline)/4
    /*margin: 5px;*/
}
```
3. Install the xblock into the lms and cms. Restart both services.
4. Now Open the installed xblock in the lms or cms and find out the value changed in the second step.
5. We can also hit following command in console to verify the variable
```shell
getComputedStyle(document.documentElement).getPropertyValue('--baseline')
```
